File: app/routes/context.py
from __future__ import annotations

from fastapi import APIRouter, Depends

from app.dependencies import get_retriever, get_store
from app.models import (
    ContextRetrieveRequest,
    ContextRetrieveResponse,
    FactRecord,
    ObservationRecord,
    PlaybookRecord,
    StoreFactRequest,
    StoreObservationRequest,
    StorePlaybookRequest,
    stable_identifier,
)
from app.retrieval import ContextRetriever
from app.store import ContextStore

router = APIRouter(prefix="/v1/context", tags=["context"])


def _fact_from_request(request: StoreFactRequest) -> FactRecord:
    payload = request.model_dump(exclude={"id"})
    return FactRecord(id=request.id or stable_identifier("fact", payload), **payload)


def _playbook_from_request(request: StorePlaybookRequest) -> PlaybookRecord:
    payload = request.model_dump(exclude={"id"})
    return PlaybookRecord(id=request.id or stable_identifier("playbook", payload), **payload)


def _observation_from_request(request: StoreObservationRequest) -> ObservationRecord:
    payload = request.model_dump(exclude={"id"})
    return ObservationRecord(id=request.id or stable_identifier("observation", payload), **payload)


@router.post("/retrieve", response_model=ContextRetrieveResponse)
def retrieve_context(
    request: ContextRetrieveRequest,
    retriever: ContextRetriever = Depends(get_retriever),
) -> ContextRetrieveResponse:
    return retriever.retrieve(request)


@router.post("/facts", response_model=FactRecord)
def store_fact(
    request: StoreFactRequest,
    store: ContextStore = Depends(get_store),
) -> FactRecord:
    return store.put_fact(_fact_from_request(request))


@router.post("/playbooks", response_model=PlaybookRecord)
def store_playbook(
    request: StorePlaybookRequest,
    store: ContextStore = Depends(get_store),
) -> PlaybookRecord:
    return store.put_playbook(_playbook_from_request(request))


@router.post("/observations", response_model=ObservationRecord)
def store_observation(
    request: StoreObservationRequest,
    store: ContextStore = Depends(get_store),
) -> ObservationRecord:
    return store.put_observation(_observation_from_request(request))

File: src/routes/context.rs
use axum::extract::State;
use axum::routing::post;
use axum::{Json, Router};

use crate::error::AppError;
use crate::models::*;
use crate::routes::AppState;
use crate::store::Store;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/v1/context/retrieve", post(retrieve_context))
        .route("/v1/context/facts", post(store_fact))
        .route("/v1/context/playbooks", post(store_playbook))
        .route("/v1/context/observations", post(store_observation))
}

async fn retrieve_context(
    State(retriever): State<AppState>,
    Json(request): Json<ContextRetrieveRequest>,
) -> Result<Json<ContextRetrieveResponse>, AppError> {
    let response = retriever.retrieve(&request).await?;
    Ok(Json(response))
}

async fn store_fact(
    State(retriever): State<AppState>,
    Json(request): Json<StoreFactRequest>,
) -> Result<Json<FactRecord>, AppError> {
    let record = request.into_record();
    let stored = retriever.store().put_fact(record).await?;
    Ok(Json(stored))
}

async fn store_playbook(
    State(retriever): State<AppState>,
    Json(request): Json<StorePlaybookRequest>,
) -> Result<Json<PlaybookRecord>, AppError> {
    let record = request.into_record();
    let stored = retriever.store().put_playbook(record).await?;
    Ok(Json(stored))
}

async fn store_observation(
    State(retriever): State<AppState>,
    Json(request): Json<StoreObservationRequest>,
) -> Result<Json<ObservationRecord>, AppError> {
    let record = request.into_record();
    let stored = retriever.store().put_observation(record).await?;
    Ok(Json(stored))
}
